nav2_behavior_tree/plugins/action/truncate_path_local_action.cpp

```diff
@@ -0,0 +1,178 @@
+  }
+
+  // expand forwards to extract desired length
+  auto forward_pose_it = nav2_util::geometry_utils::first_after_integrated_distance(
+    current_pose, path_.poses.end(), distance_forward);
+
+  // expand backwards to extract desired length
+  // Note: current_pose + 1 is used because reverse iterator points to a cell before it
+  auto backward_pose_it = nav2_util::geometry_utils::first_after_integrated_distance(
+    std::reverse_iterator(current_pose + 1), path_.poses.rend(), distance_backward);
+
+  nav_msgs::msg::Path output_path;
+  output_path.header = path_.header;
+  output_path.poses = std::vector<geometry_msgs::msg::PoseStamped>(
+    backward_pose_it.base(), forward_pose_it);
+  setOutput("output_path", output_path);
+
+  return BT::NodeStatus::SUCCESS;
+}
+
+inline void TruncatePathLocal::updatePath(bool reset_path_pruning_on_change)
+{
+  if (reset_path_pruning_on_change) {
+    nav_msgs::msg::Path new_path;
+    getInput("input_path", new_path);
+
+    // detect for changes
+    bool changed = true;
+    if (new_path.poses.size() == path_.poses.size()) {
+      changed = false;
+      for (size_t i = 0; i < new_path.poses.size(); i++) {
+        if (path_.poses[i].pose != new_path.poses[i].pose) {
+          changed = true;
+          break;
+        }
+      }
+    }
+
+    if (changed) {
+      path_ = new_path;
+      closest_pose_detection_begin_ = path_.poses.begin();
+    }
+  } else {
+    getInput("input_path", path_);
+    closest_pose_detection_begin_ = path_.poses.begin();
+  }
+}
+
+inline bool TruncatePathLocal::getRobotPose(
+  std::string path_frame_id, geometry_msgs::msg::PoseStamped & pose)
+{
+  if (!getInput("pose", pose)) {
+    std::string robot_frame;
+    if (!getInput("robot_frame", robot_frame)) {
+      RCLCPP_ERROR(
+        config().blackboard->get<rclcpp::Node::SharedPtr>("node")->get_logger(),
+        "Neither pose nor robot_frame specified for %s", name().c_str());
+      return false;
+    }
+    double transform_tolerance;
+    getInput("transform_tolerance", transform_tolerance);
+    if (!nav2_util::getCurrentPose(
+        pose, *tf_buffer_, path_frame_id, robot_frame, transform_tolerance))
+    {
+      RCLCPP_WARN(
+        config().blackboard->get<rclcpp::Node::SharedPtr>("node")->get_logger(),
+        "Failed to lookup current robot pose for %s", name().c_str());
+      return false;
+    }
+  }
+  return true;
+}
+
+double
+TruncatePathLocal::poseDistance(
+  const geometry_msgs::msg::PoseStamped & pose1,
+  const geometry_msgs::msg::PoseStamped & pose2,
+  const double angular_distance_weight)
+{
+  double dx = pose1.pose.position.x - pose2.pose.position.x;
+  double dy = pose1.pose.position.y - pose2.pose.position.y;
+  tf2::Quaternion q1;
+  tf2::convert(pose1.pose.orientation, q1);
+  tf2::Quaternion q2;
+  tf2::convert(pose2.pose.orientation, q2);
+  double da = angular_distance_weight * std::abs(q1.angleShortestPath(q2));
+  return std::sqrt(dx * dx + dy * dy + da * da);
```

Comment:
Yes, it can be still helpful. It can help e.g. picking the correct pose near cusps (where robot often finds itself spatially closer to segment _after_ the cusp, but still is not rotated enough, so technically it is still _before_ the cusp) or it can be used for loop problems if user decides (due to API elegancy flaws/statefulness risk reasons I stated in comments above) to use the non-stateful approach.

Reply:
OK, an inline comment explaining the value would be  great. Else 10 years from now someone will file this as a bug and I want an explicit mention that this is to deal with orientation changes due to looping / cusps in the paths